conan/internal/rest/pkg_sign.py

```diff
@@ -12,34 +15,53 @@ def __init__(self, cache, home_folder):
-            self._plugin_verify_function = mod.verify
+            self._plugin_sign_function = getattr(mod, "sign", None)
+            self._plugin_verify_function = getattr(mod, "verify", None)
         else:
             self._plugin_sign_function = self._plugin_verify_function = None
 
+    @property
+    def is_sign_configured(self):
+        return self._plugin_sign_function is not None
+
+    @property
+    def is_verify_configured(self):
+        return self._plugin_verify_function is not None
+
+    def sign_pkg(self, ref, files, folder):
+        metadata_sign = os.path.join(folder, METADATA, "sign")
+        mkdir(metadata_sign)
+        # Generate the package sign manifest before calling the plugin
+        _save_manifest(folder, metadata_sign)
+        signatures = self._plugin_sign_function(ref, artifacts_folder=folder,
+                                                signature_folder=metadata_sign)
+        # Save signatures file with the plugin's returned signatures data
+        _save_signatures(metadata_sign, signatures)
+        # Add files to package bundle so they get uploaded
+        files[f"{METADATA}/sign/{PKGSIGN_MANIFEST}"] = get_manifest_filepath(metadata_sign)
+        files[f"{METADATA}/sign/{PKGSIGN_SIGNATURES}"] = get_signatures_filepath(metadata_sign)
+        for sig in signatures:
+            for name, file in sig.get("sign_artifacts", {}).items():
+                #TODO: print output?
+                files[f"{METADATA}/sign/{file}"] = os.path.join(metadata_sign, file)
+
     def sign(self, upload_data):
-        if self._plugin_sign_function is None:
+        if not self.is_sign_configured:
             return
 
-        def _sign(ref, files, folder):
-            metadata_sign = os.path.join(folder, METADATA, "sign")
-            mkdir(metadata_sign)
-            self._plugin_sign_function(ref, artifacts_folder=folder, signature_folder=metadata_sign)
-            for f in os.listdir(metadata_sign):
-                files[f"{METADATA}/sign/{f}"] = os.path.join(metadata_sign, f)
-
         for rref, packages in upload_data.items():
             recipe_bundle = upload_data.recipe_dict(rref)
             if recipe_bundle["upload"]:
-                _sign(rref, recipe_bundle["files"], self._cache.recipe_layout(rref).download_export())
+                self.sign_pkg(rref, recipe_bundle["files"],
+                              self._cache.recipe_layout(rref).download_export())
             for pref in packages:
                 pkg_bundle = upload_data.package_dict(pref)
                 if pkg_bundle["upload"]:
-                    _sign(pref, pkg_bundle["files"], self._cache.pkg_layout(pref).download_package())
+                    self.sign_pkg(pref, pkg_bundle["files"],
+                                  self._cache.pkg_layout(pref).download_package())
 
     def verify(self, ref, folder, files):
-        if self._plugin_verify_function is None:
+        if not self.is_verify_configured:
```

Comment:
Same as above, this should be an assert (if any, it can be just removed), not a no-op ``if``